src/Stack/Nix.hs

```diff
@@ -0,0 +1,180 @@
+        exePath <- liftIO getExecutablePath
+        return (exePath, args)
+
+-- | If Nix is enabled, re-runs the OS command returned by the second argument in a
+-- Nix container.  Otherwise, runs the inner action.
+--
+-- This takes an optional release action just like `reexecWithOptionalShell`.
+execWithOptionalShell
+    :: M env m
+    => m (FilePath,[String])
+    -> IO ()
+    -> m ()
+execWithOptionalShell getCmdArgs inner =
+  do config <- asks getConfig
+     inShell <- getInShell
+     isReExec <- asks getReExec
+     if | inShell && not isReExec ->
+            throwM OnlyOnHostException
+        | inShell ->
+            liftIO (do inner
+                       exitSuccess)
+        | not (nixEnable (configNix config)) ->
+            do liftIO inner
+               liftIO exitSuccess
+        | otherwise ->
+            do runShellAndExit
+                 getCmdArgs
+
+runShellAndExit :: M env m
+                => m (String, [String])
+                -> m ()
+runShellAndExit getCmdArgs = do
+     config <- asks getConfig
+     envOverride <- getEnvOverride (configPlatform config)
+     (cmnd,args) <- getCmdArgs
+     isStdoutTerminal <- asks getTerminal
+     (isStdinTerminal,isStderrTerminal) <-
+       liftIO ((,) <$> hIsTerminalDevice stdin
+                   <*> hIsTerminalDevice stderr)
+     let mshellFile = nixInitFile (configNix config)
+         pkgsInConfig = nixPackages (configNix config)
+     if not (null pkgsInConfig) && isJust mshellFile then
+       throwM NixCannotUseShellFileAndPackagesException
+       else return ()
+     let isTerm = isStdinTerminal && isStdoutTerminal && isStderrTerminal
+         nixopts = case mshellFile of
+           Just filePath -> [filePath]
+           Nothing -> ["-E", intercalate " " $ concat
+                              [["with (import <nixpkgs> {});"
+                               ,"runCommand \"myEnv\" {"
+                               ,"buildInputs=lib.optional stdenv.isLinux glibcLocales ++ lib.optional stdenv.isDarwin darwin.cf-private ++ ["],pkgsInConfig,["];"
+                               ,"shellHook=''"
+                               ,   "STACK_IN_NIX_EXTRA_ARGS='"]
+                               ,      (map (\p -> concat ["--extra-lib-dirs=", "${"++p++"}/lib"
+                                                         ," --extra-include-dirs=", "${"++p++"}/include"])
+                                           pkgsInConfig), ["' ;"
+                               ,"'';"
+                               ,"} \"\""]]]
+                    -- glibcLocales is necessary on Linux to avoid warnings about GHC being incapable to set the locale.
+         fullArgs = concat [ -- ["--pure"],
+                            map T.unpack (nixShellOptions (configNix config))
+                           ,nixopts
+                           ,["--command", ("export " ++ inShellEnvVar ++ "=1 ; ")
+                                          ++ intercalate " " (cmnd:args)
```

Comment:
This looks like it may not handle arguments with spaces or shell-interpreted characters properly.